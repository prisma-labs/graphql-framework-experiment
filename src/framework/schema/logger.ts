import { rootLogger } from '../../utils/logger'
export const log = rootLogger.child('schema')
