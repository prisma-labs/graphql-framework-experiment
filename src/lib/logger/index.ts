export { create, RootLogger as Logger } from './main'
