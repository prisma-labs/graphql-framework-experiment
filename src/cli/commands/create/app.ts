import { stripIndent } from 'common-tags'
import * as fs from 'fs-jetpack'
import * as Path from 'path'
import prompts from 'prompts'
import { PackageJson } from 'type-fest'
import { Command } from '../../../lib/cli'
import { rightOrFatal } from '../../../lib/utils'
import * as Layout from '../../../lib/layout'
import { tsconfigTemplate } from '../../../lib/layout/tsconfig'
import { rootLogger } from '../../../lib/nexus-logger'
import { ownPackage } from '../../../lib/own-package'
import * as PackageManager from '../../../lib/package-manager'
import * as PluginRuntime from '../../../lib/plugin'
import * as PluginWorktime from '../../../lib/plugin/worktime'
import * as proc from '../../../lib/process'
import { createGitRepository, CWDProjectNameOrGenerate } from '../../../lib/utils'

const log = rootLogger.child('cli').child('create').child('app')

const SQLITE_DEFAULT_CONNECTION_URI = 'file:./dev.db'

export default class App implements Command {
  async parse() {
    await run({})
  }
}

interface ConfigInput {
  projectName: string
}

interface InternalConfig {
  projectName: string
  projectRoot: string
  sourceRoot: string
}

/**
 * TODO
 */
export async function run(configInput?: Partial<ConfigInput>): Promise<void> {
  if (process.env.NEXUS_CREATE_HANDOFF === 'true') {
    await runLocalHandOff()
  } else {
    await runBootstrapper(configInput)
  }
}

/**
 * TODO
 */
export async function runLocalHandOff(): Promise<void> {
  log.trace('start local handoff')

  const parentData = await loadDataFromParentProcess()
  const layout = rightOrFatal(await Layout.create())
  const pluginM = await PluginWorktime.getUsedPlugins(layout)
  const plugins = PluginRuntime.importAndLoadWorktimePlugins(pluginM, layout)
  log.trace('plugins', { plugins })

  // TODO select a template

  for (const p of plugins) {
    await p.hooks.create.onAfterBaseSetup?.({
      database: parentData.database,
      connectionURI: parentData.connectionURI,
    })
  }
}

/**
 * TODO
 */
export async function runBootstrapper(configInput?: Partial<ConfigInput>): Promise<void> {
  log.trace('start bootstrapper')

  await assertIsCleanSlate()

  const projectName = configInput?.projectName ?? CWDProjectNameOrGenerate()
  const internalConfig: InternalConfig = {
    projectName: projectName,
    projectRoot: process.cwd(),
    sourceRoot: Path.join(process.cwd(), 'api'),
    ...configInput,
  }
  const nexusVersion = getNexusVersion()
  const packageManager = await getPackageManager(internalConfig.projectRoot)

  if (packageManager === 'sigtermed') {
    return
  }

  const database = await getDatabaseSelection()

  if (database === 'sigtermed') {
    return
  }

  // TODO in the future scan npm registry for nexus plugins, organize by
  // github stars, and so on.

  log.info('Scaffolding project')

  await scaffoldBaseFiles(internalConfig)

  log.info(`Installing dependencies`, {
    nexusVersion,
  })

  await packageManager.addDeps([`nexus@${nexusVersion}`], {
    require: true,
  })

  //
  // install additional deps
  //

  const deps = []
  const addDepsConfig: PackageManager.AddDepsOptions = {
    envAdditions: {},
  }

  // [1]
  // This allows installing prisma without a warning being emitted about there
  // being a missing prisma schema. For more detail refer to
  // https://prisma-company.slack.com/archives/CEYCG2MCN/p1575480721184700 and
  // https://github.com/prisma/photonjs/blob/master/packages/photon/scripts/generate.js#L67

  if (database) {
    deps.push(`nexus-plugin-prisma@${getPrismaPluginVersion()}`)
    addDepsConfig.envAdditions!.SKIP_GENERATE = 'true' // 1
    saveDataForChildProcess({
      database: database.choice,
      connectionURI: database.connectionURI,
    })
    // Allow the plugin to be enabled so that nexus can run the `onAfterBaseSetup` hook
    await scaffoldTemplate(templates.prisma(internalConfig))
  } else {
    await scaffoldTemplate(templates.helloWorld(internalConfig))
  }

  if (deps.length) {
    await packageManager.addDeps(deps, {
      ...addDepsConfig,
      require: true,
    })
  }

  await packageManager.addDeps(['prettier'], {
    require: true,
    dev: true,
  })

  //
  // pass off to local create
  //

  await packageManager
    .runBin('nexus create', {
      stdio: 'inherit',
      envAdditions: { NEXUS_CREATE_HANDOFF: 'true' },
      require: true,
    })
    .catch((error) => {
      console.error(error.message)
      process.exit(error.exitCode ?? 1)
    })

  //
  // return to global create
  //

  // Any disk changes after this point will show up as dirty working directory
  // for the user
  await createGitRepository()

  // If the user setup a db driver but not the connection URI yet, then do not
  // enter dev mode yet. Dev mode will result in a runtime-crashing app.
  if (!(database && !database.connectionURI)) {
    // We will enter dev mode with the local version of nexus. This is a kind
    // of cheat, but what we want users to have as their mental model. When they
    // terminate this dev session, they will restart it typically with e.g. `$
    // yarn dev`. This global-nexus-process-wrapping-local-nexus-process
    // is unique to bootstrapping situations.
    log.info('Starting dev mode')

    await packageManager
      .runScript('dev', {
        stdio: 'inherit',
        envAdditions: { NEXUS_CREATE_HANDOFF: 'true' },
        require: true,
      })
      .catch((error) => {
        console.error(error.message)
        process.exit(error.exitCode ?? 1)
      })
  }
}

async function getPackageManager(projectRoot: string): Promise<PackageManager.PackageManager | 'sigtermed'> {
  const packageManagerTypeEnvVar = process.env.CREATE_APP_CHOICE_PACKAGE_MANAGER_TYPE as
    | PackageManager.PackageManagerType
    | undefined

  const packageManagerType = packageManagerTypeEnvVar ?? (await askForPackageManager())

  if (packageManagerType === 'sigtermed') {
    return 'sigtermed'
  }

  const packageManager = PackageManager.createPackageManager(packageManagerType, { projectRoot })

  return packageManager
}

function getNexusVersion(): string {
  const nexusVersionEnvVar = process.env.CREATE_APP_CHOICE_NEXUS_VERSION
  const nexusVersion = nexusVersionEnvVar ?? `${ownPackage.version}`
  return nexusVersion
}

async function getDatabaseSelection(): Promise<DatabaseSelection | 'sigtermed'> {
  const envar = process.env.CREATE_APP_CHOICE_DATABASE_TYPE as Database | 'NO_DATABASE' | undefined

  if (envar) {
    if (envar === 'NO_DATABASE') {
      return null
    }
    if (envar === 'SQLite') {
      return {
        database: true,
        choice: envar,
        connectionURI: SQLITE_DEFAULT_CONNECTION_URI,
      }
    }
    return {
      database: true,
      choice: envar,
      connectionURI: undefined,
    }
  }

  return await askForDatabase()
}

// TODO this data should come from db driver

export type Database = 'SQLite' | 'PostgreSQL' | 'MySQL'

type DatabaseSelection =
  | null
  | 'sigtermed'
  | {
      database: true
      choice: Database
      connectionURI: string | undefined
    }

/**
 * Ask the user if they would like to use a database driver.
 */
async function askForDatabase(): Promise<DatabaseSelection> {
  let {
    usePrisma,
  }: {
    usePrisma?: boolean
  } = await prompts({
    type: 'confirm',
    name: 'usePrisma',
    message: 'Do you want to use a database? (https://prisma.io)',
  })

  if (usePrisma === undefined) {
    return 'sigtermed'
  }

  if (usePrisma === false) {
    return null
  }
  // TODO the supported databases should come from the plugin driver...
  let { database }: { database?: Database } = await prompts({
    type: 'select',
    name: 'database',
    message: 'Choose a database',
    choices: [
      {
        title: 'PostgreSQL',
        description: 'Requires running a PostgreSQL database',
        value: 'PostgreSQL',
      },
      {
        title: 'SQLite',
        description: 'No operational overhead',
        value: 'SQLite',
      },
      {
        title: 'MySQL',
        description: 'Requires running a MySQL database',
        value: 'MySQL',
      },
    ] as any, // Typings are missing the 'description' property...
    initial: 0,
  })

  if (database === undefined) {
    return 'sigtermed'
  }

  if (database === 'SQLite') {
    return {
      database: true,
      choice: 'SQLite',
      connectionURI: SQLITE_DEFAULT_CONNECTION_URI,
    }
  }

  return { database: true, choice: database, connectionURI: undefined }
}

/**
 * Ask the user if they would like to use npm or yarn.
 * TODO if we detect that yarn is installed on the user's machine then we should
 * default to that, otherwise default to npm.
 */
async function askForPackageManager(): Promise<PackageManager.PackageManagerType | 'sigtermed'> {
  const choices: {
    title: string
    value: PackageManager.PackageManagerType
  }[] = [
    { title: 'yarn', value: 'yarn' },
    { title: 'npm', value: 'npm' },
  ]

  type Result = {
    packageManagerType?: PackageManager.PackageManagerType
  }

  const result: Result = await prompts({
    name: 'packageManagerType',
    type: 'select',
    message: 'What is your preferred package manager?',
    choices,
  })

  if (result.packageManagerType === undefined) {
    return 'sigtermed'
  }

  return result.packageManagerType
}

/**
 * Check that the cwd is a suitable place to start a new nexus project.
 */
async function assertIsCleanSlate() {
  log.trace('checking folder is in a clean state')
  const contents = await fs.listAsync()

  if (contents !== undefined && contents.length > 0) {
    proc.fatal('Cannot create a new nexus project here because the directory is not empty', {
      contents,
    })
  }
}

async function scaffoldTemplate(template: Template) {
  return Promise.all(
    template.files.map(({ path, content }) => {
      return fs.writeAsync(path, content)
    })
  )
}

interface TemplateCreator {
  (internalConfig: InternalConfig): Template
}

interface Template {
  files: { path: string; content: string }[]
}

type TemplateName = 'helloWorld' | 'prisma'

const templates: Record<TemplateName, TemplateCreator> = {
  helloWorld(internalConfig) {
    return {
      files: [
        {
          path: Path.join(internalConfig.sourceRoot, 'graphql.ts'),
          content: stripIndent`
            import { schema } from "nexus";
      
            schema.addToContext(req => {
              return {
                memoryDB: {
                  worlds: [
                    { id: "1", population: 6_000_000, name: "Earth" },
                    { id: "2", population: 0, name: "Mars" }
                  ]
                }
              }
            })
      
            schema.objectType({
              name: "World",
              definition(t) {
                t.id("id")
                t.string("name")
                t.float("population")
              }
            })
      
            schema.queryType({
              definition(t) {        
                t.field("hello", {
                  type: "World",
                  args: {
                    world: schema.stringArg({ required: false })
                  },
                  resolve(_root, args, ctx) {
                    const worldToFindByName = args.world ?? "Earth"
                    const world = ctx.memoryDB.worlds.find(w => w.name === worldToFindByName)
      
                    if (!world) throw new Error(\`No such world named "\${args.world}"\`)
      
                    return world
                  }
                })
      
                t.list.field('worlds', {
                  type: 'World',
                  resolve(_root, _args, ctx) {
                    return ctx.memoryDB.worlds
                  } 
                })
              }
            })
          `,
        },
      ],
    }
  },
  prisma(internalConfig) {
    return {
      files: [
        {
          path: Path.join(internalConfig.sourceRoot, 'app.ts'),
          content: stripIndent`
          import { use } from 'nexus'
          import { prisma } from 'nexus-plugin-prisma'
          
          use(prisma())
        `,
        },
      ],
    }
  },
}

/**
 * Scaffold a new nexus project from scratch
 */
async function scaffoldBaseFiles(options: InternalConfig) {
  const appEntrypointPath = Path.join(options.sourceRoot, 'app.ts')
  const sourceRootRelative = Path.relative(options.projectRoot, options.sourceRoot)

  await Promise.all([
    // Empty app and graphql module.
    // Having at least one of these satisfies minimum Nexus requirements.
    // We put both to setup vscode debugger config with an entrypoint that is
    // unlikely to change.
    fs.writeAsync(
      appEntrypointPath,
      stripIndent`
      /**
       * This file is your server entrypoint. Don't worry about its emptyness, Nexus handles everything for you.
       * However, if you need to add settings, enable plugins, schema middleware etc, this is place to do it.
       * Below are some examples of what you can do. Uncomment them to try them out!
       */

      /**
       * Change a variety of settings
       */

      // import { settings } from 'nexus'
      //
      // settings.change({
      //   server: {
      //     port: 4001
      //   }
      // })

      /**
       * Add some schema middleware
       */

      // import { schema } from 'nexus'
      //
      // schema.middleware((_config) => {
      //   return async (root, args, ctx, info, next) {
      //     ctx.log.trace('before resolver')
      //     await next(root, args, ctx, info)
      //     ctx.log.trace('after resolver')
      //   }
      // })

      /**
       * Enable the Prisma plugin. (Needs \`nexus-plugin-prisma\` installed)
       */

      // import { use } from 'nexus'
      // import { prisma } from 'nexus-plugin-prisma'
      //
      // use(prisma())
    `
    ),
    fs.writeAsync(Path.join(options.sourceRoot, 'graphql.ts'), ''),
    // An exhaustive .gitignore tailored for Node can be found here:
    // https://github.com/github/gitignore/blob/master/Node.gitignore
    // We intentionally stay minimal here, as we want the default ignore file
    // to be as meaningful for nexus users as possible.
    fs.writeAsync(
      '.gitignore',
      stripIndent`
        # Node
        node_modules
        npm-debug.log*
        yarn-debug.log*
        yarn-error.log*
        lerna-debug.log*
      `
    ),
    fs.writeAsync(Path.join(options.projectRoot, 'package.json'), {
      name: options.projectName,
      license: 'UNLICENSED',
      version: '0.0.0',
      dependencies: {},
      scripts: {
        format: "npx prettier --write './**/*.{ts,md}'",
        dev: 'nexus dev',
        build: 'nexus build',
        start: `node .nexus/build/${sourceRootRelative}`,
      },
      prettier: {
        semi: false,
        singleQuote: true,
        trailingComma: 'all',
      },
    } as PackageJson),

    // todo should be supplied by the plugins
    fs.writeAsync('.prettierignore', './prisma/**/*.md'),

    fs.writeAsync(
      'tsconfig.json',
      tsconfigTemplate({
        sourceRootRelative,
        outRootRelative: Layout.DEFAULT_BUILD_DIR_PATH_RELATIVE_TO_PROJECT_ROOT,
      })
    ),

    fs.writeAsync(
      '.vscode/launch.json',
      stripIndent`
        {
          // Note: You can delete this file if you're not using vscode
          "version": "0.2.0",
          "configurations": [
            {
              "type": "node",
              "request": "launch",
              "name": "Debug Nexus App",
              "protocol": "inspector",
              "runtimeExecutable": "\${workspaceRoot}/node_modules/.bin/nexus",
              "runtimeArgs": ["dev"],
              "args": ["${Path.relative(options.projectRoot, appEntrypointPath)}"],
              "sourceMaps": true,
              "console": "integratedTerminal"
            }
          ]
        }    
      `
    ),
  ])
}

const ENV_PARENT_DATA = 'NEXUS_CREATE_DATA'

type ParentData = {
  database?: PluginRuntime.OnAfterBaseSetupLens['database']
  connectionURI?: PluginRuntime.OnAfterBaseSetupLens['connectionURI']
}

async function loadDataFromParentProcess(): Promise<ParentData> {
  if (process.env[ENV_PARENT_DATA]) {
    const data = JSON.parse(process.env[ENV_PARENT_DATA]!)
    log.trace('loaded parent data', { data })
    return data
  }
  log.trace('no parent data found to load')
  return {}
}

function saveDataForChildProcess(data: ParentData): void {
  process.env[ENV_PARENT_DATA] = JSON.stringify(data)
}

/**
 * Helper function for fetching the correct version of prisma plugin to
 * install. Useful for development where we can override the version installed
 * by environment variable NEXUS_PLUGIN_PRISMA_VERSION.
 */
function getPrismaPluginVersion(): string {
  let prismaPluginVersion: string
  if (process.env.NEXUS_PLUGIN_PRISMA_VERSION) {
    log.warn(
      'found NEXUS_PLUGIN_PRISMA_VERSION defined. This is only expected if you are actively developing on nexus right now',
      {
        NEXUS_PLUGIN_PRISMA_VERSION: process.env.NEXUS_PLUGIN_PRISMA_VERSION,
      }
    )
    prismaPluginVersion = process.env.NEXUS_PLUGIN_PRISMA_VERSION
  } else {
    prismaPluginVersion = 'latest'
  }
  return prismaPluginVersion
}
