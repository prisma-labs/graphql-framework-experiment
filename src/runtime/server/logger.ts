import { rootLogger } from '../../lib/nexus-logger'

export const log = rootLogger.child('server')
