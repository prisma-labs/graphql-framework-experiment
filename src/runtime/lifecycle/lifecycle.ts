import * as NexusSchema from '@nexus/schema'
import { rootLogger } from '../../lib/nexus-logger'
import { AppState } from '../app'

const log = rootLogger.child('lifecycle')

/**
 * The data we pass to callbacks
 */
type Data = {
  schema: NexusSchema.core.NexusGraphQLSchema
}

/**
 * The function users can register
 */
type Callback = (data: Data) => void

export type LazyState = {
  callbacks: {
    start: Callback[]
  }
}

export function createLazyState(): LazyState {
  return {
    callbacks: {
      start: [],
    },
  }
}

/**
 * Public component interface
 */
export interface Lifecycle {
  /**
   * Register callback to be run when the application starts.
   *
   * @remarks
   *
   * Put initialization code here that you don't want run during [Nexus reflection](https://nxs.li/about/reflection).
   */
  start(callback: (data: Data) => void): void
  // idea - if we ever need it
  // phase: {
  //   now: 'reflection' | 'runtime'
  //   reflection: boolean
  //   runtime: boolean
  // }
  // todo
  // stage: {
  //   now: 'production' | 'development'
  // }
  // todo - requires having calls be stripped at build time for tree-shaking
  // reflection: {}
}

/**
 * Internal component controls
 */
export interface Private {
  reset(): void
  trigger: {
    start(data: Data): void
  }
}

/**
 * Control the Lifecycle component
 */
export interface Controller {
  public: Lifecycle
  private: Private
}

/**
 * Create an instance of Lifecycle
 */
export function create(state: AppState): Controller {
  state.components.lifecycle = createLazyState()

  const api = {} as Lifecycle

  api.start = function (callback) {
    log.debug('registered callback', { event: 'start' })
    state.components.lifecycle.callbacks.start.push(callback)
  }

  return {
    public: api,
    private: {
      reset() {
        state.components.lifecycle.callbacks.start.length = 0
      },
      trigger: {
        start(data) {
          for (const callback of state.components.lifecycle.callbacks.start) {
            log.debug('will run callback', { event: 'start' })
            try {
              callback(data)
            } catch (error) {
              const wrappedError = new Error(`Lifecycle callback error on event "start":\n\n${error.message}`)
              wrappedError.stack = error.stack
              throw wrappedError
            }
            log.debug('did run callback', { event: 'start' })
          }
        },
      },
    },
  }
}
