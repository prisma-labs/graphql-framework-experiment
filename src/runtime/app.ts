import * as Logger from '@nexus/logger'
import * as NexusSchema from '@nexus/schema'
import { rootLogger } from '../lib/nexus-logger'
import * as Plugin from '../lib/plugin'
import { RuntimeContributions } from '../lib/plugin'
import * as Reflection from '../lib/reflection/stage'
import { builtinScalars } from '../lib/scalars'
import { Index } from '../lib/utils'
import * as Lifecycle from './lifecycle'
import * as Schema from './schema'
import * as Server from './server'
import * as Settings from './settings'
import { assertAppNotAssembled } from './utils'

const log = Logger.log.child('app')

// WARNING Make sure that jsdoc edits here are ported to runtime/index

export interface App {
  /**
   * [API Reference](https://nxs.li/docs/api/logger) ⌁ [Guide](https://nxs.li/docs/guides/logger) ⌁ [Issues](https://nxs.li/issues/components/logger)
   */
  log: Logger.Logger
  /**
   * [API Reference](https://nxs.li/docs/api/server) ⌁ [Guide](https://nxs.li/docs/guides/server) ⌁ [Issues](https://nxs.li/issues/components/server)
   */
  server: Server.Server
  /**
   * [API Reference](https://nxs.li/docs/api/schema) ⌁ [Guide](https://nxs.li/docs/guides/schema) ⌁ [Issues](https://nxs.li/issues/components/schema)
   */
  schema: Schema.Schema
  /**
   * [API Reference](https://nxs.li/docs/api/settings) ⌁ [Issues](https://nxs.li/issues/components/settings)
   */
  settings: Settings.Settings
  /**
   * [API Reference](https://nxs.li/docs/api/on) ⌁ [Issues](https://nxs.li/issues/components/lifecycle)
   *
   * Use the lifecycle component to tap into application events.
   */
  on: Lifecycle.Lifecycle
  /**
   * [API Reference](https://nxs.li/docs/api/use-plugins) ⌁ [Issues](https://nxs.li/issues/components/plugins)
   */
  use(plugin: Plugin.Plugin): void
  /**
   * Run this to gather the final state of all Nexus api interactions. This method
   * is experimental. It provides experimental support for NextJS integration.
   *
   * In a regular Nexus app, you should not need to use this method.
   *
   * @experimental
   */
  assemble(): any
  /**
   * This method makes it possible to reset the state of the singleton. This can
   * be useful when working in a development environment where multiple runs of
   * the app (or run-like, e.g. Node module cache being reset) can take place
   * without having state reset. Such an example of that is the Next.js dev
   * mode.
   *
   * @experimental
   */
  reset(): any
  /**
   * todo
   */
  start(): any
  /**
   * todo
   */
  stop(): any
  /**
   * todo
   */
}

export type AppState = {
  plugins: Plugin.Plugin[]
  /**
   * Once the app is started incremental component APIs can no longer be used. This
   * flag let's those APIs detect that they are being used after app start. Then
   * they can do something useful like tell the user about their mistake.
   */
  assembled: null | {
    settings: Settings.SettingsData
    schema: NexusSchema.core.NexusGraphQLSchema
    missingTypes: Index<NexusSchema.core.MissingType>
    loadedPlugins: RuntimeContributions<any>[]
    createContext: Schema.ContextAdder
  }
  running: boolean
  components: {
    schema: Schema.LazyState
    lifecycle: Lifecycle.LazyState
  }
}

export type PrivateApp = App & {
  private: {
    state: AppState
  }
}

/**
 * Create new app state. Be careful to pass this state to components to complete its
 * data. The data returned only contains core state, despite what the return
 * type says.
 */
export function createAppState(): AppState {
  const appState: AppState = {
    assembled: null,
    running: false,
    plugins: [],
    components: {} as any, // populated by components
  }

  return appState
}

/**
 * Create an app instance
 */
export function create(): App {
  const state = createAppState()
  const serverComponent = Server.create(state)
  const schemaComponent = Schema.create(state)
  const settingsComponent = Settings.create(state, {
    serverSettings: serverComponent.private.settings,
    schemaSettings: schemaComponent.private.settings,
    log: Logger.log,
  })
  const lifecycleComponent = Lifecycle.create(state)

  const app: App = {
    log: log,
    settings: settingsComponent.public,
    schema: schemaComponent.public,
    server: serverComponent.public,
    on: lifecycleComponent.public,
    reset() {
      rootLogger.debug('resetting state')
      schemaComponent.private.reset()
      serverComponent.private.reset()
      settingsComponent.private.reset()
      lifecycleComponent.private.reset()
      state.assembled = null
      state.plugins = []
      state.running = false
      dogfood()
    },
    async start() {
      if (Reflection.isReflection()) return
      if (state.running) return
      if (!state.assembled) {
        throw new Error('Must call app.assemble before calling app.start')
      }
      lifecycleComponent.private.trigger.start({
        schema: state.assembled!.schema,
      })
      await serverComponent.private.start()
      state.running = true
    },
    async stop() {
      if (Reflection.isReflection()) return
      if (!state.running) return
      await serverComponent.private.stop()
      state.running = false
    },
    use(plugin) {
      assertAppNotAssembled(state, 'app.use', 'The plugin you attempted to use will be ignored')
      state.plugins.push(plugin)
    },
    assemble() {
      if (state.assembled) return

      schemaComponent.private.beforeAssembly()

      /**
       * Plugin reflection is run in the same process (eval). This means if the
       * process is the app, which it is during testing for example, then we
       * need to take extreme care to not mark assembly as complete, during
       * plugin reflection. If we did, then, when we would try to start the app,
       * it would think it is already assembled. !
       */
      if (Reflection.isReflectionStage('plugin')) return

      state.assembled = {} as AppState['assembled']

      const loadedPlugins = Plugin.importAndLoadRuntimePlugins(state.plugins, state.components.schema.scalars)
      state.assembled!.loadedPlugins = loadedPlugins

      const { schema, missingTypes } = schemaComponent.private.assemble(loadedPlugins)
      state.assembled!.schema = schema
      state.assembled!.missingTypes = missingTypes

      if (Reflection.isReflectionStage('typegen')) return

      const { createContext } = serverComponent.private.assemble(loadedPlugins, schema)
      state.assembled!.createContext = createContext

      const { settings } = settingsComponent.private.assemble()
      state.assembled!.settings = settings

      schemaComponent.private.checks()
    },
  }

  /**
   * Dogfood the public API to change things.
   */
  function dogfood() {
    /**
     * Setup default log filter
     */
    app.settings.change({
      logger: {
        filter: 'app:*, nexus:*@info+, *@warn+',
        pretty: { timeDiff: false },
      },
    })

    /**
     * Setup default scalar types
     */
    app.schema.importType(builtinScalars.DateTime, 'date')
    app.schema.importType(builtinScalars.Json, 'json')

    /**
     * Add `req` and `res` to the context by default
     */
    app.schema.addToContext((params) => params)
  }

  // HACK dogfood function called eagerly here once and
  // then within reset method. We should have a better
  // reset system.

  dogfood()

  return {
    ...app,
    private: {
      state: state,
    },
  } as App
}
