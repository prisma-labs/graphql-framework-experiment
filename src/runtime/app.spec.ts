import { log } from '@nexus/logger'
import * as GraphQL from 'graphql'
import * as HTTP from 'http'
import 'jest-extended'
import * as Lo from 'lodash'
import { inspect } from 'util'
import { setReflectionStage, unsetReflectionStage } from '../lib/reflection'
import * as App from './app'
import * as Lifecycle from './lifecycle'

let app: App.PrivateApp

function dump(x: any) {
  return inspect(x, { depth: null })
}

beforeEach(() => {
  app = App.create() as App.PrivateApp
  // fix once there is a singleton logger
  log.settings({ pretty: { timeDiff: false } })
  app.settings.change({ logger: { pretty: { color: false, timeDiff: false } } })
})

describe('reset', () => {
  it('reset the app state', () => {
    const originalAppState = Lo.cloneDeep(app.private.state)
    app.settings.change({
      server: { path: '/bar' },
      schema: { connections: { foo: {} } },
    })
    app.schema.objectType({
      name: 'Foo',
      definition(t) {
        t.string('ok')
      },
    })
    app.on.start(() => {})
    app.assemble()
    app.reset()
    expect(app.settings.current.server.path).toEqual(app.settings.original.server.path)
    expect(app.settings.current.schema).toEqual(app.settings.original.schema)
    // must dump because functions are not equal-able
    expect(dump(app.private.state)).toEqual(dump(originalAppState))
    expect(app.private.state.components.lifecycle).toEqual(Lifecycle.createLazyState())
  })

  it('calling before assemble is fine', () => {
    app.reset()
  })

  it('multiple calls is effectively a noop', () => {
    app.reset()
    app.reset()
  })
})

describe('assemble', () => {
  const spy = createLogSpy()

  beforeEach(() => {
    // avoid schema check error
    app.schema.objectType({
      name: 'Foo',
      definition(t) {
        t.string('ok')
      },
    })
  })

  it('multiple calls is a noop', () => {
    app.assemble()
    app.assemble()
  })

  describe('warnings when api features used after assembly', () => {
    it('settings.change', () => {
      app.assemble()
      app.settings.change({ server: { path: '/foo' } })
      expect(spy.log.mock.calls).toMatchSnapshot()
    })

    it('schema.use', () => {
      app.assemble()
      app.schema.use({ config: { name: 'foo' } })
      expect(spy.log.mock.calls).toMatchSnapshot()
    })

    // todo all api methods
  })
})

describe('lifecycle', () => {
  beforeEach(() => {
    app.settings.change({ server: { port: 7583 } })
    app.schema.queryType({
      definition(t) {
        t.string('foo')
      },
    })
  })
  afterEach(async () => {
    await app.stop()
  })
  describe('start', () => {
    it('callback is called with data when app is started', async () => {
      const fn = jest.fn()
      app.on.start(fn)
      app.assemble()
      await app.start()
      expect(fn.mock.calls[0][0].schema instanceof GraphQL.GraphQLSchema).toBeTrue()
    })
    it('if callback throws error then Nexus shows a nice error', async () => {
      app.on.start(() => {
        throw new Error('error from user code')
      })
      app.assemble()
      expect(await app.start().catch((e: Error) => e)).toMatchInlineSnapshot(`
        [Error: Lifecycle callback error on event "start":

        error from user code]
      `)
    })
  })
})

describe('checks', () => {
  const spy = createLogSpy()

  it('if graphql schema is empty upon assemble then there is a warning', () => {
    app.assemble()
    expect(spy.log.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "â–² nexus:schema Your GraphQL schema is empty. This is normal if you have not defined any GraphQL types yet. If you did however, check that your files are contained in the same directory specified in the \`rootDir\` property of your tsconfig.json file.
      ",
        ],
      ]
    `)
  })
})

describe('server', () => {
  it('has raw.http to get access to underling node http server', () => {
    expect(app.server.raw.http).toBeInstanceOf(HTTP.Server)
  })

  describe('handlers', () => {
    it('under reflection are noops', () => {
      setReflectionStage('plugin')
      const g = app.server.handlers.graphql as any
      expect(g()).toBeUndefined()
      unsetReflectionStage()
    })

    // todo, process exit poop
    it.todo('if accessed before assembly, and not under reflection, error')
  })
})

/**
 * helpers
 */

function createLogSpy() {
  const spy = {} as { log: jest.SpyInstance }

  beforeEach(() => {
    spy.log = jest.spyOn(process.stdout, 'write').mockImplementation(() => {
      return true
    })
  })

  afterEach(() => {
    spy.log.mockRestore()
  })

  return spy
}
