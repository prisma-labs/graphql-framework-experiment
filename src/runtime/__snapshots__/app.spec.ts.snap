// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assemble warnings when api features used after assembly schema.use 1`] = `
Array [
  Array [
    "â–² nexus Cannot call [33mapp.schema.use[39m(...) after app.assemble()

The Nexus Schema plugin you used will be ignored.
",
  ],
]
`;

exports[`assemble warnings when api features used after assembly settings.change 1`] = `
Array [
  Array [
    "â–² nexus Cannot call [33mapp.settings.change[39m(...) after app.assemble()

Your change of settings will be ignored.
",
  ],
]
`;
